speakerbox/main.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from pathlib import Path
+from typing import Dict, Optional, Union
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from datasets import Audio, Dataset, DatasetDict, arrow_dataset, load_metric
+from sklearn.metrics import ConfusionMatrixDisplay
+from transformers import (
+    EvalPrediction,
+    Trainer,
+    TrainingArguments,
+    Wav2Vec2FeatureExtractor,
+    Wav2Vec2ForSequenceClassification,
+    feature_extraction_utils,
+    pipeline,
+)
+
+###############################################################################
+
+DEFAULT_BASE_MODEL = "superb/wav2vec2-base-superb-sid"
+
+###############################################################################
+
+
+def train(
+    dataset: Union[
+        pd.DataFrame,
+        Dataset,
+        DatasetDict,
+    ],
+    model_name: str = "trained-speakerbox",
+    model_base: str = DEFAULT_BASE_MODEL,
+    max_duration: float = 2.0,
+    batch_size: int = 2,
+) -> Path:
+    """
+    Train a speaker classification model.
+
+    Parameters
+    ----------
+    dataset: Union[pd.DataFrame, Dataset, DatasetDict]
+        The dataset to use for training.
+
+        Should only contain the columns/features: "label" and "audio".
+
+        If provided a pandas DataFrame, this will convert it into a Dataset object
+        and cast the "label" column into ClassLabels and the "audio" column into Audio.
+        If provided a Dataset object (or the converted pandas DataFrame),
+        this will convert it into a DatasetDict with
+        random train, test, and validation splits.
+    model_name: str
+        A name for the model. This will also create a directory with the same name
+        to store the produced model in.
+        Default: "trained-speakerbox"
+    model_base: str
+        The model base to use before fine tuning.
+    max_duration: float
+        The maximum duration to use for each audio clip.
+        Any clips longer than this will be trimmed.
+        Default: 2.0
+    batch_size: int
+        The number of examples to use in a batch during training.
+        Default: 2
+
+    Returns
+    -------
+    model_storage_path: Path
+        The path to the directory where the model is stored.
+    """
+    # Load feature extractor
+    feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(model_base)
+
+    # Convert provided dataset to HF Dataset
+    if isinstance(dataset, pd.DataFrame):
+        dataset = Dataset.from_pandas(dataset, preserve_index=False)
+        dataset = dataset.class_encode_column("label")
+        dataset = dataset.cast_column("audio", Audio(feature_extractor.sampling_rate))
+
+    # Convert into train, test, and validate dict
+    if isinstance(dataset, Dataset):
+        train_and_test = dataset.train_test_split(test_size=0.4)
+        test_and_valid = train_and_test["test"].train_test_split(test_size=0.5)
```

Comment:
Are the frequencies of the speaker labels about the same across the train, test, and validation set?

Could we not intermix events across these three sets? That is, audio files from an event should belong only to one set. I think this would better simulate the real-world use of the model, after training.

Reply:
Yes as part of the setup / data prep process I am making sure that there are an even number of samples for each speaker label and then doing random samples for the train, test, and validation sets. So while there may be some randomness in how many of each speaker is in each set, the overall dataset is being reduced to where they have an even number and we would have to be incredibly unlucky to get really unbalanced subsets.

I really like the idea of having a couple of events that arent in the training set in the holdout validation / balancing the dataset by both speaker and event. Unfortunately our dataset is only made up of five events in total 😬 

A part of me wants to say I should run this model across of a couple of transcripts to preseed annotations and then manually run through and fix the annotations where it got them wrong for quickly adding more training data.

Reply:
Here is where I am doing the speaker balancing: https://github.com/CouncilDataProject/speakerbox/blob/feature/dataset-expansion/speakerbox/ds/utils.py#L195